### PR TITLE
impl(generator): handle http path with multiple substitutions

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #10980 

When dealing with protos generated from discovery documents, we can encounter http paths with multiple substitutions, i.e. multiple pairs of `{}`. We need to parse this and then reassemble it with the string literals pieces and appropriate request accessor calls.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502)

### CODE REVIEW

File: generator/internal/http_option_utils_test.cc [hunk 1]
```diff
@@ -324,6 +354,24 @@ TEST_F(HttpOptionUtilsTest, SetHttpDerivedMethodVarsExtensionInfo) {
...
+      service_file_descriptor->service(0)->method(5);
+  VarsDictionary vars;
+  SetHttpDerivedMethodVars(ParseHttpExtension(*method), *method, vars);
+  EXPECT_THAT(vars.at("method_request_url_path"),
+              Eq("/v1/projects/{project=project}/instances/{instance=instance}/"
+                 "databases"));
+  EXPECT_THAT(vars.at("method_request_body"), Eq("*"));
+  EXPECT_THAT(vars.at("method_http_verb"), Eq("Post"));
+  EXPECT_THAT(vars.at("method_rest_path"),
+              Eq("absl::StrCat(\"/v1/projects/\", request.project(), "
```
Comment (coryan) [inline generator/internal/http_option_utils_test.cc:371]: nit: Maybe a raw string would make that match more readable.

File: generator/internal/http_option_utils.cc [hunk 2]
```diff
@@ -242,6 +255,20 @@ ParseHttpExtension(google::protobuf::MethodDescriptor const& method) {
...
   info.path_suffix = match[4];
+
+  std::size_t current = 0;
+  for (auto open = url_pattern.find('{'); open != std::string::npos;
+       open = url_pattern.find('{', current)) {
+    info.rest_path.emplace_back(url_pattern.substr(current, open - current),
+                                false);
+    current = open + 1;
+    auto close = url_pattern.find('}', current);
+    std::pair<std::string, std::string> foo = absl::StrSplit(
```
Comment (coryan) [inline generator/internal/http_option_utils.cc:266]: `foo`?

File: generator/internal/http_option_utils.h [hunk 3]
```diff
@@ -31,13 +31,20 @@ struct HttpSimpleInfo {
 };
 
 struct HttpExtensionInfo {
+  struct RestPathPiece {
+    RestPathPiece(std::string piece, bool format_as_accessor)
+        : piece(std::move(piece)), format_as_accessor(format_as_accessor) {}
+    std::string piece;
+    bool format_as_accessor;
```
Comment (coryan) [inline generator/internal/http_option_utils.h:38]: This smells like a type code in disguise.  Maybe want you want is:

```
using RestPathPiece = std::function<std::string(google::protobuf::MethodDescriptor const&)>;
...
std::vector<RestPath> rest_path;
```

and then you just initialize them to either:

```
info.rest_path.emplace_back([piece = foo.first](google::protobuf::MethodDescriptor const& method) {
     return absl::StrFormat("request.%s()", FormatFieldAccessorCall(method, p.piece));
});
```

or to:

```
info.rest_path.emplace_back([piece = foo.first](google::protobuf::MethodDescriptor const&) {
     return absl::StrFormat("\"%s\"", p.piece);
});
```

**Review by coryan — CHANGES_REQUESTED** [review]
Some ideas for cleaning up things.  The use of meta-syntactic variables is not fantastic.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @coryan)

___
*[`generator/internal/http_option_utils.h` line 38 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#gh-1185554237:-NUcrKbC8pLPAYLnDSO7:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/844e8abe9b6a94386ed55d5c1f5794ed3647b5f2/generator/internal/http_option_utils.h#L38)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This smells like a type code in disguise.  Maybe want you want is:

```
using RestPathPiece = std::function<std::string(google::protobuf::MethodDescriptor const&)>;
...
std::vector<RestPath> rest_path;
```

and then you just initialize them to either:

```
info.rest_path.emplace_back([piece = foo.first](google::protobuf::MethodDescriptor const& method) {
     return absl::StrFormat("request.%s()", FormatFieldAccessorCall(method, p.piece));
});
```

or to:

```
info.rest_path.emplace_back([piece = foo.first](google::protobuf::MethodDescriptor const&) {
     return absl::StrFormat("\"%s\"", p.piece);
});
```

</blockquote></details>

Done.
___
*[`generator/internal/http_option_utils.cc` line 266 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#gh-1185549900:-NUcrUvv0GAmDljSGZwP:b-1l9j4a) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/844e8abe9b6a94386ed55d5c1f5794ed3647b5f2/generator/internal/http_option_utils.cc#L266)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

`foo`?
</blockquote></details>

Oops, changed.
___
*[`generator/internal/http_option_utils_test.cc` line 371 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#gh-1185547390:-NUcrXZZ-PMHv4jpOSTs:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/844e8abe9b6a94386ed55d5c1f5794ed3647b5f2/generator/internal/http_option_utils_test.cc#L371)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: Maybe a raw string would make that match more readable.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502)* status: all files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`generator/internal/http_option_utils.cc` line 264 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#-NUcw99q62FeVgORLxvu:-NUcw99rDnlZeuFdKFBO:bvkpmrw) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bb08c5c554a52119545254d22ea228ff4725c6b7/generator/internal/http_option_utils.cc#L264)):*
> ```cpp
>     auto close = url_pattern.find('}', current);
>     std::pair<std::string, std::string> param = absl::StrSplit(
>         url_pattern.substr(current, close - current), absl::ByChar('='));
> ```

Do you need to limit the number of splits, or this happens automatically when the return type is `std::pair<>`?  Just want to learn

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`generator/internal/http_option_utils.cc` line 264 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#-NUcw99q62FeVgORLxvu:-NUcz_t51OjU1gtvRxr7:bp61qo3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bb08c5c554a52119545254d22ea228ff4725c6b7/generator/internal/http_option_utils.cc#L264)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Do you need to limit the number of splits, or this happens automatically when the return type is `std::pair<>`?  Just want to learn 
</blockquote></details>

When using `std::pair<>` it will discard any additional splits it finds. So it's an implicit limit of 2.

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan and @scotthart)

___
*[`generator/internal/http_option_utils.cc` line 264 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#-NUcw99q62FeVgORLxvu:-NUczkv_EV9vaLW_vzfI:bqm2jqy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/bb08c5c554a52119545254d22ea228ff4725c6b7/generator/internal/http_option_utils.cc#L264)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

When using `std::pair<>` it will discard any additional splits it finds. So it's an implicit limit of 2.
</blockquote></details>

When splitting to a pair you'll get the first two and the rest will be discarded.

**Comment by coryan** [comment]
> When using `std::pair<>` it will discard any additional splits it finds. So it's an implicit limit of 2.
> 

A limit usually means that a=b=c split by = returns a and b=c... Do you really want just a and b?

-- sent from my phone.

**Review by scotthart — COMMENTED** [review]
Yes. Some foundational assumptions of the input we're ingesting would be violated if there was more than `{a=b}` present between curly braces.

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11502#-:-NUd8ZtU2eVFDiQREVom:b-np9sr4)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

**Pull request merged**